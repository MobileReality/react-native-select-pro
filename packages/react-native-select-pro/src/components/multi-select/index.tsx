import React, { ComponentPropsWithRef } from 'react';
import {
    ScrollView,
    StyleSheet,
    useWindowDimensions,
    ViewStyle,
} from 'react-native';

import { parsePercentageToNumber } from '../../helpers';
import type { OptionalToRequired } from '../../helpers/types/OptionalToRequired';
import type {
    OnPressSelectControlType,
    OnSetPosition,
    OptionType,
    Select,
} from '../../index';
import type { DispatchType, State } from '../../state/types';
import { SelectInput } from '../select-input';

import { MultiSelectedOption } from './components/selected-option';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'selectControlTextStyle'
    | 'placeholderText'
    | 'selectControlStyle'
    | 'disabled'
    | 'searchPattern'
    | 'searchable'
    | 'multiSelection'
    | 'multiSelectionOptionStyle'
    | 'placeholderTextColor'
>;

type SelectControlProps = OptionalToRequired<
    FromSelectComponentProps & Pick<State, 'selectedOption'>
>;

type Props = {
    onPressRemove: (option: OptionType | null) => void;
    onPressSelectControl: OnPressSelectControlType;
} & SelectControlProps &
    Pick<State, 'isOpened' | 'selectedOption' | 'searchValue'> & {
        dispatch: DispatchType;
    } & { setPosition: OnSetPosition };

export const MultiSelect = ({
    searchable,
    selectControlTextStyle,
    selectControlStyle,
    selectedOption,
    placeholderText,
    placeholderTextColor,
    onPressRemove,
    disabled,
    dispatch,
    isOpened,
    onPressSelectControl,
    searchPattern,
    searchValue,
    setPosition,
    multiSelection,
    multiSelectionOptionStyle,
}: Props) => {
    const { width } = useWindowDimensions();
    const selectedOptionTyped = selectedOption as OptionType[];

    const resolveSelectedOptions = () => {
        if (!selectedOptionTyped) {
            if (searchable) {
                return null;
            }
            return (
                <MultiSelectedOption
                    isPlaceholder={true}
                    option={null}
                    optionWidth="100%"
                    placeholderText={placeholderText}
                    placeholderTextColor={placeholderTextColor}
                    selectControlTextStyle={selectControlTextStyle}
                />
            );
        }

        const optionWidth = () => {
            const WIDTH_THRESHOLD = 100;
            const WIDTH_OFFSET = 72;
            const { length } = selectedOptionTyped;
            const initialWidth = selectControlStyle
                ? (selectControlStyle as ViewStyle).width
                : 100;
            let calculatedWidth = 100;
            if (typeof initialWidth === 'number') {
                calculatedWidth = (initialWidth - WIDTH_OFFSET) / length;
                if (calculatedWidth < WIDTH_THRESHOLD) {
                    return WIDTH_THRESHOLD;
                }
                return Math.floor(calculatedWidth);
            }
            if (typeof initialWidth === 'string') {
                const ratioToScreen = Math.floor(
                    width * (parsePercentageToNumber(initialWidth) / 100),
                );
                calculatedWidth = ratioToScreen / length;
                if (calculatedWidth - WIDTH_OFFSET < WIDTH_THRESHOLD) {
                    return WIDTH_THRESHOLD;
                }
                return calculatedWidth - WIDTH_OFFSET;
            }
            return 0;
        };

        return selectedOptionTyped.map((option: OptionType, index) => {
            return (
                <MultiSelectedOption
                    key={index}
                    option={option}
                    optionWidth={optionWidth()}
                    placeholderText={placeholderText}
                    selectControlTextStyle={selectControlTextStyle}
                    multiSelectionOptionStyle={multiSelectionOptionStyle}
                    onPressRemove={onPressRemove}
                />
            );
        });
    };

    return (
        <ScrollView horizontal={true} style={styles.multiSelectionWrapper}>
            {searchable && (
                <SelectInput
                    disabled={disabled}
                    dispatch={dispatch}
                    isOpened={isOpened}
                    multiSelection={multiSelection}
                    placeholderText={placeholderText}
                    placeholderTextColor={placeholderTextColor}
                    searchPattern={searchPattern}
                    searchValue={searchValue}
                    selectControlTextStyle={selectControlTextStyle}
                    selectedOption={selectedOption}
                    setPosition={setPosition}
                    onPressSelectControl={onPressSelectControl}
                />
            )}
            {resolveSelectedOptions()}
        </ScrollView>
    );
};

type Styles = {
    multiSelectionWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    multiSelectionWrapper: {
        flex: 1,
    },
});
