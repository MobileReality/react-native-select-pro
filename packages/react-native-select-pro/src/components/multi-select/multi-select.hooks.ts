import { useCallback, useMemo } from 'react';
import { StyleSheet, useWindowDimensions } from 'react-native';

import { useSelectContext } from '../../context';
import { dimensionPercentageToDP, getReducedSectionData } from '../../helpers';
import type { OnPressRemove, OptionType } from '../../types';
import { isOptionIndexType, isOptionType, isSectionOptionsType } from '../../types';

import type { UseMultiSelectProps } from './multi-select.types';

const WIDTH_THRESHOLD = 100;
const WIDTH_OFFSET = 72;

export const useMultiSelect = <T>({ selectedOptions }: UseMultiSelectProps<T>) => {
    const { width: screenWidth } = useWindowDimensions();

    const {
        dispatch,
        searchValue,
        optionsData,
        selectedOptionIndex,
        disabled,
        selectedOption,
        multiple,
        onRemove,
        styles,
    } = useSelectContext();

    const { select: selectStyles } = styles ?? {};
    const { multiSelectedOption: multiSelectedCustomStyles } = selectStyles ?? {};

    const removeSingleOption = useCallback(() => {
        dispatch({
            type: 'selectOption',
            payload: {
                selectedOption: null,
                selectedOptionIndex: -1,
            },
        });

        const isSearchable = typeof searchValue === 'string';
        if (isSearchable) {
            dispatch({
                type: 'setSearchValue',
                payload: '',
            });
        }
    }, [dispatch, searchValue]);

    const removeOptionInMultipleSelect = useCallback(
        (option: OptionType<T>, selectedOptions: OptionType<T>[]) => {
            const removedSelectedOptions = selectedOptions.filter(
                (selected) => selected.value !== option.value,
            );

            const isSectionedOptions = isSectionOptionsType(optionsData);
            const resolvedOptionsData = isSectionedOptions
                ? getReducedSectionData(optionsData)
                : optionsData;
            const foundIndex = resolvedOptionsData.findIndex(
                (item) => 'value' in item && item.value === option?.value,
            );

            let resolveSelectedOptionsIndexes: number | number[] = -1;
            if (Array.isArray(selectedOptionIndex)) {
                const filteredIndexes = selectedOptionIndex.filter((item) => item !== foundIndex);
                resolveSelectedOptionsIndexes =
                    filteredIndexes.length > 0 ? filteredIndexes : resolveSelectedOptionsIndexes;
            }

            dispatch({
                type: 'selectOption',
                payload: {
                    selectedOption:
                        removedSelectedOptions.length > 0 ? removedSelectedOptions : null,
                    selectedOptionIndex: resolveSelectedOptionsIndexes,
                },
            });

            return { index: foundIndex, option };
        },
        [dispatch, optionsData, selectedOptionIndex],
    );

    const onPressRemove: OnPressRemove<T> = useCallback(
        (option = null) => {
            if (disabled) {
                return;
            }
            let removedOption;
            if (option && multiple && selectedOptions) {
                removedOption = removeOptionInMultipleSelect(option, selectedOptions);
            } else {
                removeSingleOption();
                removedOption = {
                    option: selectedOption,
                    index: selectedOptionIndex,
                };
            }

            if (
                onRemove &&
                removedOption.option &&
                isOptionType(removedOption.option) &&
                isOptionIndexType(removedOption.index)
            ) {
                // callback
                onRemove(removedOption.option, removedOption.index);
            }
        },
        [
            disabled,
            multiple,
            onRemove,
            removeOptionInMultipleSelect,
            removeSingleOption,
            selectedOption,
            selectedOptionIndex,
            selectedOptions,
        ],
    );

    const containerWidth = StyleSheet.flatten(selectStyles?.container)?.width;

    const calculatedOptionWidth = useMemo(() => {
        if (!selectedOptions) {
            return 0;
        }
        const { length } = selectedOptions;
        const initialWidth = containerWidth ?? 100;
        if (typeof initialWidth === 'number') {
            const calculatedWidth = (initialWidth - WIDTH_OFFSET) / length;
            return calculatedWidth < WIDTH_THRESHOLD
                ? WIDTH_THRESHOLD
                : Math.floor(calculatedWidth);
        }
        if (typeof initialWidth === 'string') {
            const ratioToScreen = dimensionPercentageToDP(initialWidth, screenWidth);
            const calculatedWidth = ratioToScreen / length;
            return calculatedWidth - WIDTH_OFFSET < WIDTH_THRESHOLD
                ? WIDTH_THRESHOLD
                : calculatedWidth - WIDTH_OFFSET;
        }
        return 0;
    }, [selectedOptions, containerWidth, screenWidth]);

    const isSearchable = typeof searchValue === 'string';

    return {
        calculatedOptionWidth,
        onPressRemove,
        disabled,
        isSearchable,
        multiSelectedCustomStyles,
    };
};
