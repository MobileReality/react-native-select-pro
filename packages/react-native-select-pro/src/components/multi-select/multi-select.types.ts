import type { ComponentPropsWithRef } from 'react';

import type { OptionalToRequired } from '../../helpers';
import type { OnPressSelectControlType, OnSetPosition, OptionType, Select } from '../../index';
import type { DispatchType, State } from '../../state/types';
import type { SelectControlStyles } from '../../types/styles';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'placeholderText'
    | 'disabled'
    | 'searchPattern'
    | 'textInputProps'
    | 'multiSelection'
    | 'placeholderTextColor'
> &
    Pick<SelectControlStyles, 'multiSelectionOptionStyle' | 'textStyle' | 'containerStyle'>;

type SelectControlProps = OptionalToRequired<FromSelectComponentProps>;

export type MultiSelectProps = SelectControlProps &
    Pick<State, 'isOpened' | 'searchValue'> & {
        onPressRemove: (option: OptionType | null) => void;
        onPressSelectControl: OnPressSelectControlType;
        dispatch: DispatchType;
        setPosition: OnSetPosition;
        selectedOptions: OptionType[] | null;
    };
