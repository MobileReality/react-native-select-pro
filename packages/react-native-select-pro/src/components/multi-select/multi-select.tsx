import React from 'react';
import type { ViewStyle } from 'react-native';
import { ScrollView, StyleSheet } from 'react-native';

import type { OnPressRemove, OptionType } from '../../types';
import { MultiSelectedOption } from '../multi-selected-option';
import { SelectInput } from '../select-input';
import { SelectText } from '../select-text';

import { useMultiSelect } from './multi-select.hooks';
import type { MultiSelectProps } from './multi-select.types';

export const MultiSelect = <T,>({ selectedOptions }: MultiSelectProps<T>) => {
    const {
        calculatedOptionWidth,
        onPressRemove,
        multiSelectedCustomStyles,
        disabled,
        isSearchable,
    } = useMultiSelect<T>({
        selectedOptions,
    });

    const resolveSelectedOptionsList = () => {
        if (!selectedOptions) {
            return isSearchable ? null : <SelectText />;
        }

        return selectedOptions.map((option: OptionType<T>) => (
            <MultiSelectedOption
                key={`${option.section}-${option.value}`}
                optionWidth={calculatedOptionWidth}
                option={option}
                multiSelectedCustomStyles={multiSelectedCustomStyles}
                disabled={disabled}
                onPressRemove={onPressRemove as OnPressRemove<unknown>}
            />
        ));
    };

    return (
        <ScrollView scrollEnabled={!disabled} style={styles.container} horizontal>
            {isSearchable && <SelectInput />}
            {resolveSelectedOptionsList()}
        </ScrollView>
    );
};

type Styles = {
    container: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    container: {
        flex: 1,
    },
});
