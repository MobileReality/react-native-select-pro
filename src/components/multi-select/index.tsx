import React, { ComponentPropsWithRef } from 'react';
import { ScrollView, StyleSheet, useWindowDimensions, ViewStyle } from 'react-native';
import type { OptionType, Select } from '@mobile-reality/react-native-select-pro';
import type {
    OnPressSelectControlType,
    OnSetPosition,
} from '@mobile-reality/react-native-select-pro';

import type { OptionalToRequired } from '../../helpers';
import { parsePercentageToNumber } from '../../helpers/parsePercentageToNumber';
import type { State } from '../../state/types';
import type { DispatchType } from '../../state/types';
import { SelectInput } from '../select-input';

import { MultiSelectedOption } from './components/selected-option';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof Select>,
    | 'selectControlTextStyle'
    | 'placeholderText'
    | 'selectControlStyle'
    | 'disabled'
    | 'searchPattern'
    | 'searchable'
    | 'multiSelection'
>;

type SelectControlProps = OptionalToRequired<
    FromSelectComponentProps & Pick<State, 'selectedOption'>
>;

type Props = {
    onPressRemove: (option: OptionType | null) => void;
    onPressSelectControl: OnPressSelectControlType;
} & SelectControlProps &
    Pick<State, 'isOpened' | 'selectedOption' | 'searchValue'> & {
        dispatch: DispatchType;
    } & { setPosition: OnSetPosition };

export const MultiSelect = ({
    searchable,
    selectControlTextStyle,
    selectControlStyle,
    selectedOption,
    placeholderText,
    onPressRemove,
    disabled,
    dispatch,
    isOpened,
    onPressSelectControl,
    searchPattern,
    searchValue,
    setPosition,
    multiSelection,
}: Props) => {
    const { width } = useWindowDimensions();
    const selectedOptionTyped = selectedOption as OptionType[];

    const resolveSelectedOptions = () => {
        if (!selectedOptionTyped) {
            if (searchable) {
                return null;
            }
            return (
                <MultiSelectedOption
                    isPlaceholder={true}
                    option={null}
                    optionWidth={'100%'}
                    placeholderText={placeholderText}
                    selectControlTextStyle={selectControlTextStyle}
                />
            );
        }

        const optionWidth = () => {
            const WIDTH_THRESHOLD = 100;
            const WIDTH_OFFSET = 72;
            const length = selectedOptionTyped.length;
            const initialWidth = selectControlStyle ? (selectControlStyle as ViewStyle).width : 100;
            let calculatedWidth = 100;
            if (typeof initialWidth === 'number') {
                calculatedWidth = (initialWidth - WIDTH_OFFSET) / length;
                if (calculatedWidth < WIDTH_THRESHOLD) {
                    return WIDTH_THRESHOLD;
                }
                return Math.floor(calculatedWidth);
            }
            if (typeof initialWidth === 'string') {
                const ratioToScreen = Math.floor(
                    width * (parsePercentageToNumber(initialWidth) / 100),
                );
                calculatedWidth = ratioToScreen / length;
                if (calculatedWidth - WIDTH_OFFSET < WIDTH_THRESHOLD) {
                    return WIDTH_THRESHOLD;
                }
                return calculatedWidth - WIDTH_OFFSET;
            }
            return 0;
        };

        return selectedOptionTyped.map((option: OptionType, index) => {
            return (
                <MultiSelectedOption
                    key={index}
                    onPressRemove={onPressRemove}
                    option={option}
                    optionWidth={optionWidth()}
                    placeholderText={placeholderText}
                    selectControlTextStyle={selectControlTextStyle}
                />
            );
        });
    };

    return (
        <ScrollView horizontal={true} style={styles.multiSelectionWrapper}>
            {searchable && (
                <SelectInput
                    disabled={disabled}
                    dispatch={dispatch}
                    isOpened={isOpened}
                    multiSelection={multiSelection}
                    onPressSelectControl={onPressSelectControl}
                    placeholderText={placeholderText}
                    searchPattern={searchPattern}
                    searchValue={searchValue}
                    selectedOption={selectedOption}
                    setPosition={setPosition}
                />
            )}
            {resolveSelectedOptions()}
        </ScrollView>
    );
};

type Styles = {
    multiSelectionWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    multiSelectionWrapper: {
        flex: 1,
    },
});
