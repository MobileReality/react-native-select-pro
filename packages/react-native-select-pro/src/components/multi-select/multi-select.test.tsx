import React from 'react';

import { MultiSelect } from './multi-select';

describe('MultiSelect', () => {
    it('should generate MultiSelect snapshot', () => {
        const multiSelect = (
            <MultiSelect
                selectedOption={{ label: 'test', value: 'test' }}
                placeholderText={undefined}
                disabled={false}
                searchPattern={undefined}
                multiSelection={true}
                placeholderTextColor={undefined}
                searchable={false}
                textInputProps={{}}
                isOpened={false}
                searchValue=""
                dispatch={() => {}}
                setPosition={() => {}}
                containerStyle={{}}
                multiSelectionOptionStyle={{}}
                textStyle={{}}
                onPressRemove={() => {}}
                onPressSelectControl={() => {}}
            />
        );
        expect(multiSelect).toMatchSnapshot();
    });
});
