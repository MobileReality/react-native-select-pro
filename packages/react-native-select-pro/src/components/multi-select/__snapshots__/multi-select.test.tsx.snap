// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect should generate MultiSelect snapshot 1`] = `
<MultiSelect
  containerStyle={{}}
  disabled={false}
  dispatch={[Function]}
  isOpened={false}
  multiSelection={true}
  multiSelectionOptionStyle={{}}
  onPressRemove={[Function]}
  onPressSelectControl={[Function]}
  searchValue=""
  selectedOptions={
    [
      {
        "label": "test",
        "value": "test",
      },
    ]
  }
  setPosition={[Function]}
  textInputProps={{}}
  textStyle={{}}
/>
`;
